File: problems/factorial/factorial.js
// Recursive function example
exports.recursive = function factorial (number) {
  if (number < 2) { return 1; }

  return number * factorial(number - 1);
};

// Iterative solution
exports.iterative = function (number) {
  var result = 1;

  for (var i = 1; i <= number; i++) {
    result *= i;
  }

  return result;
};

// Iterative using a reverse loop
exports.iterativeReverse = function (number) {
  var result = 1;

  while (number) {
    result *= number;
    number -= 1;
  }

  return result;
};

File: problems/factorial/factorial.rb
class Integer
  # Iterative
  def factorial_simple(number)
    fat = 1
    (2..number).each { |n| fat = fat * n }
    fat
  end

  # Iterative Reverse Loop
  def factorial_reverse(number)
    (number - 1).downto(1).each { |i| n *= i }
    n
  end

  # Recursive - Ternary Operator
  def factorial_recursive(number)
    (number == 1) ? 1 : (number * factorial_recursive(number-1))
  end

  # The Ruby-iest version, using reduce function
  def factorial_reduce(number)
    (1..number).reduce(:*)
  end

  alias_method :factorial, factorial_reduce
end
